src/main/kotlin/org/pkl/lsp/ast/Extensions.kt

```diff
@@ -397,21 +399,36 @@ fun PklNode.findBySpan(line: Int, col: Int, includeTerminals: Boolean = false):
   return childHit ?: hit
 }
 
-fun PklNode.toLspURIString(): String {
-  return when (val file = containingFile) {
-    is StdlibFile -> "pkl-lsp://stdlib/${file.name}.pkl"
-    is FsFile -> file.uri.toString()
-    else -> {
-      val uri = file.uri.toString()
-      "pkl-lsp://${file.pklAuthority}/${URLEncoder.encode(uri, Charsets.UTF_8)}"
+val VirtualFile.lspUri
+  get(): URI {
+    return when (this) {
+      is StdlibFile -> URI("pkl-lsp://stdlib/${name}.pkl")
+      is FsFile -> uri
+      else -> {
+        val uri = uri.toString()
+        URI("pkl-lsp://${pklAuthority}/${URLEncoder.encode(uri, Charsets.UTF_8)}")
+      }
     }
   }
-}
 
-fun PklNode.toCommandURIString(): String {
-  val sp = beginningSpan()
-  val params = """["${toLspURIString()}",${sp.beginLine},${sp.beginCol}]"""
-  return "command:pkl.open.file?${URLEncoder.encode(params, Charsets.UTF_8)}"
+fun PklNode.getLocationUri(forDocs: Boolean): URI {
+  val useCommandLink = project.clientOptions.renderOpenFileCommandInDocs ?: false
+  val span = beginningSpan()
+  return when {
+    useCommandLink && forDocs -> {
+      val params = buildJsonArray {
+        add(containingFile.lspUri.toString())
+        add(span.beginLine)
+        add(span.beginCol)
+      }
+      URI("command", "pkl.open.file?$params", null)
```

Comment:
Have you checked this works without URLEncode? I also had problems if there was any spaces in the json, like after a `,`.

Reply:
Yeah, the second argument here is the raw path. It gets encoded when you `toString()` the URI.